testsuite/tests/tsan/waitgroup_stubs.c

```diff
@@ -0,0 +1,51 @@
+#include <assert.h>
+#include <stdatomic.h>
+#include <unistd.h>
+
+#include <caml/mlvalues.h>
+#include <caml/tsan.h>
+
+#define MAX_WAITGROUP   8
+#define SPIN_WAIT_MS    10
+
+typedef struct {
+  unsigned    limit; /* Number of threads participating to the checkpoint */
+  atomic_uint count; /* Number of threads that have reach the checkpoint */
+} waitgroup;
+
+static waitgroup waitgroups[MAX_WAITGROUP] = { 0 };
+
+static atomic_uint index = 0;
+
+CAMLno_tsan static waitgroup* wg_get(unsigned idx)
+{
+  assert(idx < MAX_WAITGROUP);
+
+  waitgroup* wg = &waitgroups[idx];
+  return wg;
+}
+
+CAMLno_tsan value wg_create(value n)
+{
+  waitgroup* wg = wg_get(index);
+
+  wg->limit = Int_val(n);
+  wg->count = 0;
+  return Val_int(index++);
+}
+
+CAMLno_tsan value wg_finish(value t)
+{
+  waitgroup* wg = wg_get(Int_val(t));
+
+  wg->count += 1;
+  return Val_unit;
+}
+
+CAMLno_tsan value wg_wait(value t)
+{
+  waitgroup* wg = wg_get(Int_val(t));
+
+  while (wg->count != wg->limit) { /* spinwait */ }
```

Comment:
Where you intending to have a `usleep(SPIN_WAIT_MS)` call in the busy loop here?

I feel a bit nervous about a pure busy loop doing no sleep or pause here. This could behave very badly on some single-core machines with a coarse-grained scheduler, and CI virtual machines could be like that.

Reply:
Ahaha, well spotted!
I actually have an earlier implementation of the spinwait that waits with a `usleep`:
```diff
diff --git a/testsuite/tests/tsan/waitgroup_stubs.c b/testsuite/tests/tsan/waitgroup_stubs.c
index 16f02363e6..7eb0ee8311 100644
--- a/testsuite/tests/tsan/waitgroup_stubs.c
+++ b/testsuite/tests/tsan/waitgroup_stubs.c
@@ -46,6 +46,10 @@ CAMLno_tsan value wg_wait(value t)
 {
   waitgroup* wg = wg_get(Int_val(t));
 
-  while (wg->count != wg->limit) { /* spinwait */ }
+  /* Always sleep at least once to cause a 'as if synchronized via sleep' */
+  do {
+    usleep(SPIN_WAIT_MS);
+  }
+  while (wg->count != wg->limit);
   return Val_unit;
 }
```
This version is a bit more tricky, hence the comment: the spinwait must always make at least one call to `usleep` so that TSan always reports `As if synchronized via sleep`:
```
diff --git a/testsuite/tests/tsan/array_elt.reference b/testsuite/tests/tsan/array_elt.reference
index 3e37b6d616..549cfd4c0d 100644
--- a/testsuite/tests/tsan/array_elt.reference
+++ b/testsuite/tests/tsan/array_elt.reference
@@ -9,6 +9,12 @@ WARNING: ThreadSanitizer: data race (pid=<implemspecific>)
     #1 camlArray_elt.entry <implemspecific> (<implemspecific>)
     #2 caml_program <implemspecific> (<implemspecific>)
 
+  As if synchronized via sleep:
+    #0 usleep <implemspecific> (<implemspecific>)
+    #1 wg_wait <implemspecific> (<implemspecific>)
+    #2 camlArray_elt.writer_<implemspecific> <implemspecific> (<implemspecific>)
+    #3 camlStdlib__Domain.body_<implemspecific> <implemspecific> (<implemspecific>)
+
   Mutex M<implemspecific> (<implemspecific>) created at:
     #0 pthread_mutex_init <implemspecific> (<implemspecific>)
     #1 caml_plat_mutex_init <implemspecific> (<implemspecific>)
```
Consider the case where a test spawns a new `domain`, the main one waits for it to finish on a `waitgroup`, depending on whether the other domain finished or not before the main one reached the `waitgroup`, it would or wouldn't `usleep` causing some flaky-ness in the test report.

I also prefer the `usleep` version, as it should be more energy efficient at the cost of slightly more complex code and references, but maybe(?) not as high-performance.

Reply:
I would rather have something that always takes a couple milliseconds than something that will blow up in our face in weird ways on some rate-limited CI machines.